rtcp-packet.go

```diff
@@ -0,0 +1,361 @@
+1       | fraction lost |       cumulative number of packets lost       |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |           extended highest sequence number received           |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |                      interarrival jitter                      |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |                         last SR (LSR)                         |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        |                   delay since last SR (DLSR)                  |
+        +=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+
+report  |                 SSRC_2 (SSRC of second source)                |
+block    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+2       :                               ...                             :
+        +=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+
+        |                  profile-specific extensions                  |
+        +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+*/
+
+
+import (
+  "crypto/aes"
+  "crypto/cipher"
+  "encoding/binary"
+  "errors"
+  // "encoding/hex"
+)
+
+type RTCPTypeClass uint8
+
+// https://tools.ietf.org/html/rfc3550#section-12.1
+const (
+  RTCPTypeSR         RTCPTypeClass = 200
+  RTCPTypeRR
+  RTCPTypeSDES
+  RTCPTypeBYE
+  RTCPTypeAPP
+)
+
+const (
+  RTCPHeaderSize = 8
+)
+
+type RTCPHeader struct {
+  buffer []byte
+}
+
+func (p *RTCPHeader) Clone() *RTCPHeader {
+	p2 := &RTCPHeader{
+		buffer: make([]byte, len(p.buffer)),
+	}
+
+	copy(p2.buffer, p.buffer)
+	return p2
+}
+
+func (p *RTCPHeader) GetRC() int {
+  rc := p.buffer[0] & 31   // 31 = 0b00011111
+  return int(rc);
+}
+
+func (p* RTCPHeader) SetPT(pt RTCPTypeClass) {
+  p.buffer[1] = byte(pt);
+}
+
+func (p* RTCPHeader) GetPT() RTCPTypeClass {
+  return RTCPTypeClass(p.buffer[1])
+}
+
+func (p* RTCPHeader) SetLength(length uint16) {
+  binary.BigEndian.PutUint16(p.buffer[2:], length)
+}
+
+func (p* RTCPHeader) GetLength() uint16 {
+  return binary.BigEndian.Uint16(p.buffer[2:])
+}
+
+func (p* RTCPHeader) GetLengthInBytes() uint32 {
+  return (uint32(p.GetLength())+1)*4
+}
+
+func (p* RTCPHeader) SetSenderSSRC(ssrc uint32) {
+  binary.BigEndian.PutUint32(p.buffer[4:], ssrc)
+}
+
+func (p* RTCPHeader) GetSenderSSRC() uint32 {
+  return binary.BigEndian.Uint32(p.buffer[4:])
+}
+
+
+type RTCPPacket struct {
+    header RTCPHeader
+    payload []byte
+}
+
+type RTCPCompoundPacket struct {
+  packets []*RTCPPacket
+  srtcpIndex uint32
+}
+
+func (p* RTCPCompoundPacket) GetESRTCPWord(e bool) []byte {
+  word := make([]byte, 4)
+  binary.BigEndian.PutUint32(word, p.srtcpIndex)
+
+  if e {
+    word[0] |= byte(128) // 128 = 0b10000000
+  }
+
+  return word
+}
+
+func gcmIV(ssrc, esrtcpWord, salt []byte) []byte {
```

Comment:
Assuming this is the algorithm described in <https://tools.ietf.org/html/rfc7714#section-9.1>, it would be better to use `index` or `srtcpIndex` than `esrtcpWord`.

Reply:
This is an SRTCP-specific method, so it should not be in the global namespace.  If you follow what was done with `RTPPacket` and make it a member of `RTCPPacket`, you can just pull these values from the packet buffer (except for salt).